core/reactor_common.c

```diff
@@ -988,75 +989,59 @@ trigger_handle_t _lf_schedule(environment_t *env, trigger_t* trigger, interval_t
             }
             // Hook the event into the list.
             found->next = e;
+            trigger->last_time = intended_tag.time;
```

Comment:
I thought that we wanted to keep track of the _last time of scheduling_, not tag of time of the last-scheduled event, because those can be all over the place. In other words, shouldn't this be as follows?
```suggestion
            trigger->last_time = lf_time_logical_elapsed();
```

Reply:
No.  The time at which lf_schedule is called is irrelevant.

Reply:
Note further that this will be used typically in physical actions, and logical time is not well defined outside the context of a reaction.

Reply:
The times at which `schedule` is invoked will be monotonically increasing. We have no assurance of that with the intended times, which could really undermine the mechanism. Generally, the problem I see is that you can schedule clusters of events that are very densely packed (i.e., with very little spacing), as long as the clusters have sufficient spacing between them and you schedule the events in an interleaving/round robin manner.

The same is true if we were to go by the times of scheduling (be it logical or physical) instead of the times of the events scheduled, however. The only way to actually achieve spacing on the logical time line is to scan the event queue and look for conflicts. As you pointed out, this is an expensive operation, however, and gives rise to tricky corner cases that need to be handled...

Exactly, what guarantees can we derive from the spacing as it is implemented now? It seems to me that the ability to specify an additional delay is what creates chaos and spoils our ability to provide meaningful guarantees. Perhaps we should not allow additional delay for actions that have a minimum spacing. We could have a keyword (`dynamic`) for actions that allow an additional delay and use the target language's type system to catch situations where the action was not declared accordingly. That way, we have strong guarantees for ordinary (e.g., `static`) actions and weaker guarantees for dynamic ones.